### DISCUSSION TITLE
Mimic vault datasource with second argument, using a file

### DISCUSSION THREAD

**Discussion by liaocanada** [hairyhenderson/gomplate#1182] [Q&A] [answered]
I have a template with some references to a vault datasource, which uses an additional argument to specify the secret path. For example: `(ds "vault" "dbpassword").value`. I want to mimic this locally using a file datasource (as I do not have access to the vault), without touching the template. 

Here's what I've tried:
**1. Create a JSON file like the one below, and use that as the data source:**
```JSON
// vault.json
{
    "dbpassword": "test"
}
```
```
gomplate -d vault=file:///some-path/vault.json
```
The error I get back for this was `Couldn't read datasource 'vault': Can't stat /some-path/vault.json/dbpassword: stat /some-path/vault.json/dbpassword: not a directory"`. 

**2. Create a directory `dir` which contains a file called `dbpassword`, and use the directory as the data source:**
```
// dbpassword
test
```
```
gomplate -d vault=file://some-path/dir
```
The error I get back for this was `error="template: ***: executing \"***\" at <\"dbpassword\">: can't evaluate field value in type string"`. It looks like `( ds "vault" "dbpassword")` resolves to the contents of the file as a string, so the `.value` then fails. 

**3. Same as 2, except I changed the `dbpassword` file contents to a JSON object, and added `?type=application/json`:**

 ```JSON
// dbpassword
{
    "value": "test"
}
```
```
gomplate -d vault=file://some-path/dir?type=application/json
```
This gives me `zsh: no matches found: vault=file:///some-path/dir?type=application/json`. It doesn't seem to work since the query component is applied on the directory and not the enclosed "dbpassword" file.

I also tried using a git datasource with "dbpassword" as the name of the repo, which did not work either. (by the way, it seems to look for the master branch by default, instead of main)

Is it possible to match this template `(ds "vault" "dbpassword").value` using a file datasource?

**Comment by liaocanada** [comment, marked as answer]
Just noticed the `zsh` at the beginning of that last error. Added quotes around the whole thing and it works now 🎉 

Basically I had to:
1. Create a file for each vault secret I referred to, with the file name matching the secret name
2. Make each of the file contents a JSON object with `value` being the key
3. Use the enclosing directory as the data source, append `?type=application/json`, and surround the whole thing in quotes

**Reply by hairyhenderson** [reply]
Glad you found the issue!
